### DISCUSSION TITLE
Can't seem to find information of how to configure clangd correctly

### DISCUSSION THREAD

**Discussion by mohabouje** [zed-industries/zed#6629] [Feature Requests]
I want to give a try to this editor. I am building a project with `xmake`. The build system generates a `compile_commands.json` in a folder in the root project called `.xmake`.

I want to configure the LSP server for C++ to set a specific path to the binary and pass the command that points to this file.

**Comment by JosephTLyons** [comment]
I went ahead and converted this into a Q&A Discussion - I'm not entirely sure myself if / how this can be set up, but maybe someone on thea team or community knows.

**Comment by Congyuwang** [comment]
I'm using `cmake` with `clangd`. It seems that I don't need to configure a path to clangd. I guess zed is using the default clangd server on MacOS which is `/usr/bin/clangd` from Xcode.

When building the cmake project, adding a `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON` will produce the `compile_commands.json` for clangd.

Also, the [llvm-clangd guide page](https://clangd.llvm.org/installation) gives some decent instructions on how to setup clangd. For example, use a `.clangd` file in your project for configuration.

But maybe I didn't answer your question as of how to set a specific path to a `clangd` binary.

**Reply by Congyuwang** [reply]
Zed doc does have a paragraph for LSP, but it's not very detailed: https://docs.zed.dev/configuration/configuring-zed#lsp

**Comment by zambetti** [comment]
I am also searching for clangd configuration documentation.

I've placed a .clangd configuration file in the root of my project that specifies include paths (this is also root of the open Zed 'project'); but it appears that the .clangd file is not being used by the language server.

••Update 1••

After some trial and error, the following .clangd in the root of my project worked for my situation, which is simpler than the original topic of this thread, but maybe it helps someone...

```
CompileFlags:
  Add:
    - "--include=/path/to/file.h"
    - "--include-directory=/path/to/directory-1"
    - "--include-directory=/path/to/directory-2"
```

••Update 2••

So now I've found a real need for the "lsp" property in settings.json...

```
  "lsp": {
      "clangd": {
          "header-insertion": "never"
      }
  },
```

I was hoping this would disable clangd's header insertion, but unfortunately it does not. Does anyone know if documentation exists for configuring the LSP?

**Reply by 413x1nkp** [reply]
Sorry for necrobumping, but I believe I must specify: **RELATIVE PATHS DON'T WORK**
Use absolute paths only for the `.clangd`

**Reply by EDEADLINK** [reply]
This is how you set clangd command-line options:

    "lsp": {
		"clangd": {
			"binary": {
				"path_lookup": true,
				"path": "clangd",
				"arguments": ["--header-insertion=never"]
			}
		}
    }

And leaving off the path and path_lookup does not work.
Yeah, this should be documented somewhere jesus christ, the ergonomics here suck.

**Reply by ArhumMK** [reply]
On windows and i had to set "path": to the exact directory of clangd down to clangd.exe for it to work (globally). setting up lsp's on windows has been a massive pain so far but its understandable i guess.

**Reply by Poldraunic** [reply]
@EDEADLINK thank you very much! This worked like a charm.

**Reply by paulross80** [reply]
Thank you!

**Reply by ryukinix** [reply]
@alexlnkp do you know any workaround for relative paths? I read that: https://github.com/clangd/clangd/issues/1038 but I didn't see anything promising to give a try.

**Reply by 413x1nkp** [reply]
> @alexlnkp do you know any workaround for relative paths? I read that: [clangd/clangd#1038](https://github.com/clangd/clangd/issues/1038) but I didn't see anything promising to give a try.

You can add a `compile_flags.txt` file with following contents into the root of your project (alongside `.clangd`)
```txt
-I
include/
```
This way, the `include` directory (in the root of project) is looked into when clangd is working, so headers in the `include` directory are seen.
Although that would mean that you don't really need `.clangd` file then, since all of the arguments can be specified there instead (one argument per line).

Make sure to restart Zed/workspace every time you change `compile_flags.txt`! The changes only take place this way.

I wish .clangd would support relative includes, but I doubt that will ever happen.

**Comment by ryukinix** [comment]
That suggestion worked. I removed the hard coded path from clangd file.
Thank you!

Em qui., 19 de dez. de 2024 07:57, Alex Murkoff ***@***.***>
escreveu:

**Comment by pgnickb** [comment]
So how do I disable autoimports in clangd? It breaks things way more often than it is useful.

**Comment by EDEADLINK** [comment]
My answer got burried in the fold so here it is more visible:

This is how you set clangd command-line options:

    "lsp": {
		"clangd": {
			"binary": {
				"path_lookup": true,
				"path": "clangd",
				"arguments": ["--header-insertion=never"]
			}
		}
    }

**Reply by IronFractal** [reply]
```
Language server error: clangd

oneshot canceled
-- stderr--

```
anyone else have this problem with the above solution?

**Reply by owlx56** [reply]
@IronFractal Maybe my solution can help you: https://github.com/zed-industries/zed/issues/19268#issuecomment-2753578911

Similar issues about LSP error should be taken seriously. Even if the network is working and the language server is downloaded, there may still be problems. `ra` usually does not complain, but `clangd` and `jdtls` often have various problems.

**Comment by pgnickb** [comment]
Is there a way to disable formatting on save? None of the language specific nor "formatter" flags seem to work. It is very annoying when it makes small formatting changes in a 10k line file of an open source project that has it's own formatting rules.

**Reply by EDEADLINK** [reply]
Unrelated, but [see this](https://github.com/zed-industries/zed/discussions/8219)

**Reply by pgnickb** [reply]
yeah, I've tried all those, which is why I decided to ask here as it feels like clang does some tidying up. Is there any way to debug this?

**Reply by pgnickb** [reply]
Ok there is also
```
"remove_trailing_whitespace_on_save": false,
```
Which I overlooked.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
